File: bank_importer/coder.py
"""
coder.py — Гүйлгээний авто ангилалын дүрмүүд

Хоёр компанид тусдаа дүрэм:
  - code_tt(desc, income, expense, month, counterparty) → (direction, code)
  - code_tr(desc, income, expense, month, counterparty) → (direction, code)

direction: 'M' = орлого ангилал, 'N' = зарлага ангилал, '' = тодорхойгүй
code:      '1.1.1', '1.2.1' гэх мэт (config.py-д тайлбар байна)

Дүрмийн эрэмбэ (дээрээс доош, эхний таарсан дүрэм хэрэглэгдэнэ):
  1. Тусгай харилцагч (Өнөрсайхан, охин компани гэх мэт)
  2. Тайлбарын түлхүүр үгс (шимтгэл, цалин, сургалт гэх мэт)
  3. Тээврийн ангилал (K2/K3/K9 карго, УБ-ДА чиглэл)
  4. Өмнөх сарын тааруулалт (огнооноос автомат)
  5. Үндсэн тээврийн зардал (default)
"""

import re


# ── TT компанийн ангилалын дүрэм ─────────────────────────────────────────

def code_tt(desc: str, income: float, expense: float,
            month: int | None = None, counterparty: str = '') -> tuple[str, str]:
    """
    Түмэн Тээх ХХК (TT + Golomt + MBank) гүйлгээний ангилал.

    Returns:
        ('M', '1.1.1') — орлого ангилал
        ('N', '1.2.1') — зарлага ангилал
        ('', '')       — ангилал тодорхойлж чадаагүй
    """
    d   = str(desc).lower()
    ctp = str(counterparty).lower()

    # Охин компани болон холбоотой харилцагчид
    _related = ['түмэн ресурс', 'түмэн тээх сити']

    if income > 0:
        # Тусгай харилцагчид
        if any(x in ctp for x in [
            'мастер фүүдс', 'премиум бьюлдинг', 'премиум иннова',
            'премиум конкрит', 'тера-экспресс', 'тера экспресс',
        ]):
            return 'M', '1.1.2'

        if 'өнөрсайхан' in ctp:
            return 'M', '5.1.3'

        if any(x in ctp for x in _related):
            return ('M', '5.1.1') if 'тооцоо' in d else ('M', '5.1.2')

        # Тайлбарын түлхүүр үгс
        if any(x in d for x in ['тооцоо', 'авлага']):
            return 'M', '1.1.2'
        if any(x in d for x in ['хүүгийн орлого', 'хүү', 'interest']):
            return 'M', '1.1.3'
        if any(x in d for x in ['буцаалт', 'буцаан']):
            return 'M', '1.1.4'
        if 'зээл' in d:
            return ('M', '5.1.3') if any(x in d + ctp for x in ['ажилтан', 'ажилч']) else ('M', '5.1.2')

        # Default орлого
        return 'M', '1.1.1'

    if expense > 0:
        # Тусгай харилцагчид
        if 'өнөрсайхан' in ctp:
            return 'N', '5.2.3'

        if any(x in ctp for x in _related):
            return ('N', '5.2.1') if 'тооцоо' in d else ('N', '5.2.2')

        # Банкны шимтгэл
        if any(x in d + ctp for x in ['шимтгэл', 'charge', 'хөтөлсний']):
            return 'N', '2.1.14'

        # Мост Мони (цалин шилжүүлгийн систем)
        if 'мост мони' in ctp or 'мост мони' in d:
            return 'N', '2.2.2'

        # МТА татварын шилжүүлэг
        if 'мта' in ctp and 'татвар' in ctp:
            if d.startswith('1251') or ';1251' in d:
                return 'N', '2.2.3'   # НӨАТ
            if d.startswith('126') or ';126' in d:
                return 'N', '2.2.1'   # ХХОАТ
            return 'N', '2.2.3'

        # Цалин / НДШ / татвар
        if 'цалин' in d:                              return 'N', '2.1.1'
        if 'нөат' in d:                               return 'N', '2.2.3'
        if any(x in d for x in ['эмндш', 'ндш']):    return 'N', '2.2.4'
        if 'томилолт' in d:                           return 'N', '2.1.3'
        if 'сургалт' in d:                            return 'N', '2.1.5'

        # Гацуурт түрээс
        if 'гацуурт' in ctp:                          return 'N', '2.1.10'
        if 'түрээс' in d:                             return 'N', '2.1.10'

        # Техник хэрэгсэл / тавилга
        if any(x in d for x in [
            'dell', 'notebook', 'laptop', 'ноутбук', 'компьютер',
            '16gb', '512 gb', 'ssd',
        ]):
            return 'N', '3.2.1'
        if any(x in d for x in ['сандал', 'шкаф', 'шүүгээ', 'тавилга', 'эд хогшил']):
            return 'N', '3.2.2'

        # Тээврийн зардал (K2/K3/K7/K9 карго, УБ-ДА чиглэл)
        is_transport = (
            ('кран' in d and 'гацуурт' not in ctp)
            or any(x in d for x in ['k2', 'k3', 'k7', 'k9', 'уб-', 'ub-', 'машин'])
        )
        if is_transport or 'өмнөх сар' in d:
            if 'өмнөх сар' in d:
                return 'N', '1.2.2'
            # Тайлбарт сарын тоо байвал шалгах
            if month:
                months_in_desc = [
                    int(m)
                    for m in re.findall(r'\b(\d{1,2})/\d{1,2}', d)
                    if 1 <= int(m) <= 12
                ]
                if months_in_desc and all(m < month for m in months_in_desc):
                    return 'N', '1.2.2'
            return 'N', '1.2.1'

        # Зээл
        if 'зээл' in d:
            return ('N', '5.2.3') if any(x in d + ctp for x in ['ажилтан', 'ажилч']) else ('N', '5.2.2')

        # Default зарлага → тээвэр
        return 'N', '1.2.1'

    return '', ''


# ── TR компанийн ангилалын дүрэм ─────────────────────────────────────────

def code_tr(desc: str, income: float, expense: float,
            month: int | None = None, counterparty: str = '') -> tuple[str, str]:
    """
    Түмэн Ресурс ХХК (TR) гүйлгээний ангилал.

    TT-тэй ижил логик, зөвхөн харилцагч нэр өөр:
    - 'түмэн тээх' → охин компани холбоосын ангилал
    """
    d   = str(desc).lower()
    ctp = str(counterparty).lower()

    if income > 0:
        if 'өнөрсайхан' in ctp:
            return 'M', '5.1.3'
        if 'түмэн тээх' in ctp:
            return ('M', '5.1.1') if 'тооцоо' in d else ('M', '5.1.2')

        if any(x in d for x in ['тооцоо', 'авлага']):
            return 'M', '1.1.2'
        if any(x in d for x in ['хүүгийн орлого', 'хүү', 'interest']):
            return 'M', '1.1.3'
        if any(x in d for x in ['буцаалт', 'буцаан']):
            return 'M', '1.1.4'
        if 'зээл' in d:
            return ('M', '5.1.3') if any(x in d + ctp for x in ['ажилтан', 'ажилч']) else ('M', '5.1.2')

        return 'M', '1.1.1'

    if expense > 0:
        if 'өнөрсайхан' in ctp:
            return 'N', '5.2.3'
        if 'түмэн тээх' in ctp:
            return ('N', '5.2.1') if 'тооцоо' in d else ('N', '5.2.2')

        if any(x in d + ctp for x in ['шимтгэл', 'charge']):
            return 'N', '2.1.14'
        if 'мост мони' in ctp or 'мост мони' in d:
            return 'N', '2.2.2'

        if 'цалин' in d:                              return 'N', '2.1.1'
        if 'нөат' in d:                               return 'N', '2.2.3'
        if any(x in d for x in ['эмндш', 'ндш']):    return 'N', '2.2.4'
        if 'томилолт' in d:                           return 'N', '2.1.3'
        if 'сургалт' in d:                            return 'N', '2.1.5'
        if 'гацуурт' in ctp:                          return 'N', '2.1.10'
        if 'түрээс' in d:                             return 'N', '2.1.10'

        if any(x in d for x in ['dell', 'notebook', 'laptop', 'ноутбук', 'компьютер']):
            return 'N', '3.2.1'

        is_transport = (
            ('кран' in d and 'гацуурт' not in ctp)
            or any(x in d for x in ['k2', 'k3', 'k7', 'k9', 'уб-', 'ub-', 'машин'])
        )
        if is_transport or 'өмнөх сар' in d:
            if 'өмнөх сар' in d:
                return 'N', '1.2.2'
            if month:
                months_in_desc = [
                    int(m)
                    for m in re.findall(r'\b(\d{1,2})/\d{1,2}', d)
                    if 1 <= int(m) <= 12
                ]
                if months_in_desc and all(m < month for m in months_in_desc):
                    return 'N', '1.2.2'
            return 'N', '1.2.1'

        if 'зээл' in d:
            return ('N', '5.2.3') if any(x in d + ctp for x in ['ажилтан', 'ажилч']) else ('N', '5.2.2')

        return 'N', '1.2.1'

    return '', ''


def apply_codes(transaction: dict, company: str) -> dict:
    """
    Гүйлгээний dict-д ангиллын кодыг нэмж буцаана.

    company: 'TT' | 'TR'
    """
    desc  = transaction.get('description', '')
    inc   = transaction.get('income')   or 0
    exp   = transaction.get('expense')  or 0
    month = transaction.get('txn_date').month if transaction.get('txn_date') else None
    ctpy  = transaction.get('counterparty', '')

    if company in ('TT', 'GM', 'MB'):
        direction, code = code_tt(desc, inc, exp, month, ctpy)
    else:  # TR
        direction, code = code_tr(desc, inc, exp, month, ctpy)

    return {
        **transaction,
        'income_code':  code if direction == 'M' else None,
        'expense_code': code if direction == 'N' else None,
    }

File: src/lib/bank-importer/coder.ts
// coder.ts — Гүйлгээний авто ангилалын дүрмүүд.
// Эх сурвалж: bank_importer/coder.py-г үнэн зөв хөрвүүлсэн.
//
// direction: 'M' = орлого ангилал, 'N' = зарлага ангилал, '' = тодорхойгүй
// code:      '1.1.1', '1.2.1' гэх мэт (config.ts-д тайлбар)
//
// Дүрмийн эрэмбэ (дээрээс доош, эхний таарсан дүрэм хэрэглэгдэнэ).
//
// ⚠️ Зарим банкны хуулга (жишээ нь Хаан банк) тайлбарыг ЛАТИНААР (romanized)
// илгээдэг тул түлхүүр үг бүрд кирилл + латин хувилбарыг хосоор нь өгнө.
// Ингэснээр "tsalin" (цалин), "tomilolt" (томилолт) гэх мэт латин бичвэр ч
// зөв ангилагдана — өмнө нь бүгд default 1.2.1 (тээвэр) руу унаж байсан.
import type { Direction, NormalizedTxn } from "./types";

// ── Түлхүүр үгс: кирилл + латин (romanized) хувилбарууд ───────────────────
const KW = {
  tootsoo: ["тооцоо", "tootsoo", "tootsoo"],
  avlaga: ["авлага", "avlaga"],
  huu: ["хүүгийн орлого", "хүү", "interest", "huugiin", "huu "],
  butsaalt: ["буцаалт", "буцаан", "butsaalt", "butsaan", "butsaah"],
  zeel: ["зээл", "zeel"],
  ajiltan: ["ажилтан", "ажилч", "ajiltan", "ajilch", "ajiltn"],
  shimtgel: ["шимтгэл", "charge", "хөтөлсний", "shimtgel", "hutulsnii"],
  mostMoni: ["мост мони", "most moni", "mostmoni"],
  tsalin: ["цалин", "tsalin", "calin", "tsaling"],
  noat: ["нөат", "noat", "noat "],
  ndsh: ["эмндш", "ндш", "emndsh", "ndsh"],
  tomilolt: ["томилолт", "tomilolt", "tomiolt"],
  surgalt: ["сургалт", "surgalt"],
  gatsuurt: ["гацуурт", "gatsuurt", "gatuurt"],
  kran: ["кран", "kran"],
  mashin: ["машин", "mashin"],
  // Холбоо, интернэт (утас, дата, интернэт үйлчилгээ)
  holboo: ["холбоо", "интернэт", "интернет", "mobikom", "mobicom", "unitel", "skytel", "gmobile", "g-mobile", "internet", " data ", "data ", "дата", "starlink", "старлинк", "утас"],
  // Бичиг хэрэг, оффис
  bichig: ["бичиг хэрэг", "бичиг", "bichig", "принтер", "printer", "toner", "тонер", "канц", "kants", "цаас", "tsaas", "хэвлэл", "hevlel"],
  // Орчуулга / бусад үйлчилгээ
  orchuulga: ["орчуулга", "orchuulga", "notariat", "нотариат"],
  // Барьцаа, дэнчин
  baritsaa: ["барьцаа", "baritsaa", "дэнчин", "denchin", "batlan"],
} as const;

const has = (d: string, keys: readonly string[]) => keys.some((k) => d.includes(k));

// desc дотор "N/M" хэлбэрийн сарын тоонуудыг олж, бүгд нь month-аас бага эсэхийг шалгана.
function allMonthsBefore(d: string, month: number): boolean {
  const monthsInDesc: number[] = [];
  const re = /\b(\d{1,2})\/\d{1,2}/g;
  let m: RegExpExecArray | null;
  while ((m = re.exec(d)) !== null) {
    const mm = parseInt(m[1], 10);
    if (mm >= 1 && mm <= 12) monthsInDesc.push(mm);
  }
  return monthsInDesc.length > 0 && monthsInDesc.every((mm) => mm < month);
}

// ── TT компанийн ангилалын дүрэм (TT + Golomt + MBank) ────────────────────
export function codeTt(
  desc: string,
  income: number,
  expense: number,
  month: number | null = null,
  counterparty = "",
): [Direction, string] {
  const d = String(desc).toLowerCase();
  const ctp = String(counterparty).toLowerCase();

  // Охин компани болон холбоотой харилцагчид
  const related = ["түмэн ресурс", "түмэн тээх сити"];

  if (income > 0) {
    // Тусгай харилцагчид
    if (
      [
        "мастер фүүдс", "премиум бьюлдинг", "премиум иннова",
        "премиум конкрит", "тера-экспресс", "тера экспресс",
      ].some((x) => ctp.includes(x))
    ) {
      return ["M", "1.1.2"];
    }
    if (ctp.includes("өнөрсайхан") || ctp.includes("onorsaihan")) return ["M", "5.1.3"];
    if (related.some((x) => ctp.includes(x))) {
      return has(d, KW.tootsoo) ? ["M", "5.1.1"] : ["M", "5.1.2"];
    }

    // Тайлбарын түлхүүр үгс
    if (has(d, KW.tootsoo) || has(d, KW.avlaga)) return ["M", "1.1.2"];
    if (has(d, KW.huu)) return ["M", "1.1.3"];
    if (has(d, KW.butsaalt)) return ["M", "1.1.4"];
    if (has(d, KW.zeel)) {
      return has(d + ctp, KW.ajiltan) ? ["M", "5.1.3"] : ["M", "5.1.2"];
    }

    // Default орлого
    return ["M", "1.1.1"];
  }

  if (expense > 0) {
    // Тусгай харилцагчид
    if (ctp.includes("өнөрсайхан") || ctp.includes("onorsaihan")) return ["N", "5.2.3"];
    if (related.some((x) => ctp.includes(x))) {
      return has(d, KW.tootsoo) ? ["N", "5.2.1"] : ["N", "5.2.2"];
    }

    // Банкны шимтгэл
    if (has(d + ctp, KW.shimtgel)) return ["N", "2.1.14"];

    // Мост Мони (цалин шилжүүлгийн систем)
    if (has(d + ctp, KW.mostMoni)) return ["N", "2.2.2"];

    // МТА татварын шилжүүлэг
    if (ctp.includes("мта") && ctp.includes("татвар")) {
      if (d.startsWith("1251") || d.includes(";1251")) return ["N", "2.2.3"]; // НӨАТ
      if (d.startsWith("126") || d.includes(";126")) return ["N", "2.2.1"]; // ХХОАТ
      return ["N", "2.2.3"];
    }

    // Цалин / НДШ / татвар
    if (has(d, KW.tsalin)) return ["N", "2.1.1"];
    if (has(d, KW.noat)) return ["N", "2.2.3"];
    if (has(d, KW.ndsh)) return ["N", "2.2.4"];
    if (has(d, KW.tomilolt)) return ["N", "2.1.3"];
    if (has(d, KW.surgalt)) return ["N", "2.1.5"];
    if (has(d, KW.holboo)) return ["N", "2.1.6"];
    if (has(d, KW.bichig)) return ["N", "2.1.7"];
    if (has(d, KW.orchuulga)) return ["N", "2.1.8"];
    if (has(d, KW.baritsaa)) return ["N", "2.1.9"];

    // Түрээс — ЗӨВХӨН Гацуурт ХХК-д төлсөн түрээс л энд орно. Бусад "түрээс"
    // (машин/кран түрээс гэх мэт) нь тээврийн төлбөр тул доорх тээврийн дүрэмд
    // унаж 1.2.1/1.2.2 болно.
    if (has(d, KW.gatsuurt) || ctp.includes("гацуурт")) return ["N", "2.1.10"];

    // Техник хэрэгсэл / тавилга
    if (
      ["dell", "notebook", "laptop", "ноутбук", "компьютер", "computer", "komputer", "kompyuter", "16gb", "512 gb", "ssd"].some((x) =>
        d.includes(x),
      )
    ) {
      return ["N", "3.2.1"];
    }
    if (["сандал", "шкаф", "шүүгээ", "тавилга", "эд хогшил", "tavilga", "sandal"].some((x) => d.includes(x))) {
      return ["N", "3.2.2"];
    }

    // Тээврийн зардал (K2/K3/K7/K9 карго, УБ-ДА чиглэл)
    const isTransport =
      (d.includes("кран") && !ctp.includes("гацуурт")) ||
      ["k2", "k3", "k7", "k9", "уб-", "ub-"].some((x) => d.includes(x)) ||
      has(d, KW.kran) || has(d, KW.mashin);
    if (isTransport || d.includes("өмнөх сар") || d.includes("umnuh sar")) {
      if (d.includes("өмнөх сар") || d.includes("umnuh sar")) return ["N", "1.2.2"];
      if (month && allMonthsBefore(d, month)) return ["N", "1.2.2"];
      return ["N", "1.2.1"];
    }

    // Зээл
    if (has(d, KW.zeel)) {
      return has(d + ctp, KW.ajiltan) ? ["N", "5.2.3"] : ["N", "5.2.2"];
    }

    // Default зарлага → тээвэр
    return ["N", "1.2.1"];
  }

  return ["", ""];
}

// ── TR компанийн ангилалын дүрэм ─────────────────────────────────────────
export function codeTr(
  desc: string,
  income: number,
  expense: number,
  month: number | null = null,
  counterparty = "",
): [Direction, string] {
  const d = String(desc).toLowerCase();
  const ctp = String(counterparty).toLowerCase();

  if (income > 0) {
    if (ctp.includes("өнөрсайхан") || ctp.includes("onorsaihan")) return ["M", "5.1.3"];
    if (ctp.includes("түмэн тээх")) {
      return has(d, KW.tootsoo) ? ["M", "5.1.1"] : ["M", "5.1.2"];
    }

    if (has(d, KW.tootsoo) || has(d, KW.avlaga)) return ["M", "1.1.2"];
    if (has(d, KW.huu)) return ["M", "1.1.3"];
    if (has(d, KW.butsaalt)) return ["M", "1.1.4"];
    if (has(d, KW.zeel)) {
      return has(d + ctp, KW.ajiltan) ? ["M", "5.1.3"] : ["M", "5.1.2"];
    }

    return ["M", "1.1.1"];
  }

  if (expense > 0) {
    if (ctp.includes("өнөрсайхан") || ctp.includes("onorsaihan")) return ["N", "5.2.3"];
    if (ctp.includes("түмэн тээх")) {
      return has(d, KW.tootsoo) ? ["N", "5.2.1"] : ["N", "5.2.2"];
    }

    if (has(d + ctp, KW.shimtgel)) return ["N", "2.1.14"];
    if (has(d + ctp, KW.mostMoni)) return ["N", "2.2.2"];

    if (has(d, KW.tsalin)) return ["N", "2.1.1"];
    if (has(d, KW.noat)) return ["N", "2.2.3"];
    if (has(d, KW.ndsh)) return ["N", "2.2.4"];
    if (has(d, KW.tomilolt)) return ["N", "2.1.3"];
    if (has(d, KW.surgalt)) return ["N", "2.1.5"];
    if (has(d, KW.holboo)) return ["N", "2.1.6"];
    if (has(d, KW.bichig)) return ["N", "2.1.7"];
    if (has(d, KW.orchuulga)) return ["N", "2.1.8"];
    if (has(d, KW.baritsaa)) return ["N", "2.1.9"];
    // Түрээс — ЗӨВХӨН Гацуурт ХХК-д төлсөн (бусад түрээс → тээвэр).
    if (has(d, KW.gatsuurt) || ctp.includes("гацуурт")) return ["N", "2.1.10"];

    if (["dell", "notebook", "laptop", "ноутбук", "компьютер", "computer", "komputer", "kompyuter"].some((x) => d.includes(x))) {
      return ["N", "3.2.1"];
    }

    const isTransport =
      (d.includes("кран") && !ctp.includes("гацуурт")) ||
      ["k2", "k3", "k7", "k9", "уб-", "ub-"].some((x) => d.includes(x)) ||
      has(d, KW.kran) || has(d, KW.mashin);
    if (isTransport || d.includes("өмнөх сар") || d.includes("umnuh sar")) {
      if (d.includes("өмнөх сар") || d.includes("umnuh sar")) return ["N", "1.2.2"];
      if (month && allMonthsBefore(d, month)) return ["N", "1.2.2"];
      return ["N", "1.2.1"];
    }

    if (has(d, KW.zeel)) {
      return has(d + ctp, KW.ajiltan) ? ["N", "5.2.3"] : ["N", "5.2.2"];
    }

    return ["N", "1.2.1"];
  }

  return ["", ""];
}

// Гүйлгээний dict-д ангиллын кодыг нэмж буцаана.
// company: 'TT' | 'GM' | 'MB' | 'HJ' → code_tt, бусад (TR) → code_tr
// HJ (Хотгор Жинст Транс) нь ерөнхий түлхүүр үгийн дүрмийг (codeTt доторх
// generic хэсэг) ашиглана; ТТ-ийн тусгай харилцагчийн дүрэм хоргүй тул үлдээв.
export function applyCodes(txn: NormalizedTxn, company: string): NormalizedTxn {
  const desc = txn.description ?? "";
  const inc = txn.income ?? 0;
  const exp = txn.expense ?? 0;
  const month = txn.txn_date ? txn.txn_date.getMonth() + 1 : null;
  const ctpy = txn.counterparty ?? "";

  const [direction, code] =
    company === "TT" || company === "GM" || company === "MB" || company === "HJ"
      ? codeTt(desc, inc, exp, month, ctpy)
      : codeTr(desc, inc, exp, month, ctpy);

  return {
    ...txn,
    income_code: direction === "M" ? code : null,
    expense_code: direction === "N" ? code : null,
  };
}
